plugins/org.python.pydev.parser/src/org/python/pydev/parser/grammar310/TreeBuilder310.java

```diff
@@ -632,76 +634,250 @@ public final SimpleNode onCloseNode(SimpleNode n, int arity) throws Exception {
+                    patternType[] pattern = null;
+                    exprType key = null;
+                    if (arity == 2) {
+                        SimpleNode popNode = stack.popNode();
+                        try {
+                            pattern = new patternType[] { (patternType) popNode };
+                        } catch (Exception e) {
+                            Log.log("Expected pattern. Found: " + popNode);
+                        }
+                    }
                     SimpleNode popNode = stack.popNode();
                     try {
-                        exprType node = (exprType) popNode;
-                        expr = new Expr(node);
+                        key = (exprType) popNode;
                     } catch (Exception e) {
                         Log.log("Expected expr. Found: " + popNode);
                     }
+                    return new MatchMapping(new exprType[] { key }, pattern, null);
                 } else {
-                    Log.log("Expected arity to be 1 here.");
+                    Log.log("Expected arity to be 1 or 2 here.");
                 }
-                return expr;
+                return null;
 
-            case JJTAS_PATTERN:
-                exprType pattern = null;
-                String name = null;
-                if (arity == 2) {
+            case JJTMAPPING_PATTERN:
+                if (arity > 0) {
+                    exprType[] keys = new exprType[arity];
+                    patternType[] patterns = new patternType[arity];
+                    MatchMapping matchMapping = null;
+                    for (int i = 0; i < arity; i++) {
+                        SimpleNode popNode = stack.popNode();
+                        try {
+                            matchMapping = (MatchMapping) popNode;
+                        } catch (Exception e) {
+                            Log.log("Expected expr. Found: " + popNode);
+                        }
+                        keys[i] = matchMapping.keys[0];
+                        patterns[i] = matchMapping.patterns[0];
+                    }
+                    return new MatchMapping(keys, patterns, null);
+                }
+                return new MatchMapping(new exprType[0], new patternType[0], null);
+
+            case JJTSEQUENCE_PATTERN:
+                if (arity > 0) {
+                    patternType[] patterns = new patternType[arity];
+                    for (int i = 0; i < arity; i++) {
+                        SimpleNode popNode = stack.popNode();
+                        try {
+                            patterns[i] = (patternType) popNode;
+                        } catch (Exception e) {
+                            Log.log("Expected pattern. Found: " + popNode);
+                        }
+                    }
+                    return new MatchSequence(patterns);
+                }
+                return new MatchSequence(new patternType[0]);
+
+            case JJTATTR:
```

Comment:
I think this rule  -- i.e.: `attr()` should be renamed to something as `match_class_attr` and shouldn't create the `MatchClass`, rather, creating the `MatchClass` should be in the `match_class` construct. 

You can create an additional node to store the info from the `match_class_attr` if you need info from it to create the `MatchClass`.

In general each rule should always return a node of the same type when possible.

Reply:
Yup. Actually, in this case we should be returning a token to store a keyword pattern, but as we don't have any token that already does that (i.e.: name + pattern), I used `MatchClass` to store and retrieve the key and pattern.

Do you think it's better to create a new ast Token to store keyword patterns or we should really stick to create a `match_class_attr` construct?